stablehlo/dialect/VhloOps.cpp

```diff
@@ -296,5 +298,101 @@ void VhloDialect::printAttribute(Attribute attr, DialectAsmPrinter& os) const {
   assert(succeeded(result));
 }
 
+///////////////////////////
+// Op Constraint Versioning
+///////////////////////////
+// These could be migrated to ODS in VhloOps.td if we figured out a better way
+// to represent this sort of constraint in tablegen.
+
+namespace {
+Type getVhloElementType(Type tensorType) {
+  if (auto ranked = tensorType.dyn_cast<RankedTensorV1Type>()) {
+    return ranked.getElementType();
+  }
+  return tensorType.cast<UnrankedTensorV1Type>().getElementType();
+}
+
+bool checkIfOperandAndResultElementTypesMatch(TypeRange operandTypes,
+                                              TypeRange resultTypes) {
+  SmallVector<Type> inputElementTypes{llvm::map_range(
+      operandTypes, [](Type t) { return getVhloElementType(t); })};
+  SmallVector<Type> resultElementTypes{llvm::map_range(
+      resultTypes, [](Type t) { return getVhloElementType(t); })};
+
+  if (llvm::any_of(
+          llvm::zip(inputElementTypes, resultElementTypes),
+          [&](auto pair) { return std::get<0>(pair) != std::get<1>(pair); }))
+    return true;
+
+  return false;
+}
+}  // namespace
+
+LogicalResult AllReduceOpV1::validateConstraint(mlir::Operation* op,
+                                                Version targetVersion) {
+  // Allow mismatched operand and result types in v0.17.0
+  if (checkIfOperandAndResultElementTypesMatch(getOperand().getType(),
```

Comment:
What are your thoughts on making a helper method like:

```
// Allow mismatched operand and result types in reduce ops in v0.17.0
LogicalResult verifyConstraint_0_17_0(mlir::Operation* op, Version targetVersion) {
  if (checkIfOperandAndResultElementTypesMatch(op->getOperandTypes(),
                                               op->getResultTypes()) &&
      targetVersion < Version(0, 17, 0))
    return failure();
  return success();
}
```

Then replacing all these validateConstraints bodies with:

```
return verifyConstraint_0_17_0(op, targetVersion);
```

Reply:
This looks good to me. 
This will allow modularization of future constraint changes with clear context. 
done